### PR TITLE
Add support for configurable SQS consumer backoff policies

### PR DESCRIPTION
**Pull request by jessemyers**
Implements the current (naive) backoff policy and an experimental exponential backoff.

### CODE REVIEW

File: microcosm_pubsub/backoff.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    @classmethod
+    def choose_backoff_policy(cls, name):
+        for subclass in cls.__subclasses__():
+            if subclass.__name__ == name:
+                return subclass
+        raise Exception("No backoff policy configured with class name: {}".format(name))
+
+
+class NaiveBackoffPolicy(BackoffPolicy):
```
Comment (jessemyers) [inline microcosm_pubsub/backoff.py]: This implements our current behavior. It either takes the default backoff or the message specific one. If neither are set, no backoff is applied and the queue's default will be used (30s?).

File: microcosm_pubsub/backoff.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    Uses a fixed timeout from the message or system default.
+
+    """
+    def compute_backoff_timeout(self, message, message_timeout):
+        backoff_timeout = message_timeout or self.visibility_timeout_seconds
+        # we can only set integer timeouts
+        return int(backoff_timeout) if backoff_timeout is not None else backoff_timeout
+
+
+class ExponentialBackoffPolicy(BackoffPolicy):
```
Comment (jessemyers) [inline microcosm_pubsub/backoff.py]: This policy does (almost) textbook exponential backoff except:
 - It's always integral (SQS limitation)
 - It never returns `0` (seems pointless, but I could be wrong)
 - It avoids returns more than 12 hours (SQS limitation)

File: microcosm_pubsub/consumer.py [hunk 3]
```diff
@@ -65,20 +67,23 @@ def nack(self, message, visibility_timeout_seconds=None):
...
         """
-        timeout = visibility_timeout_seconds or self.visibility_timeout_seconds
-        if timeout is not None:
-            self.sqs_client.change_message_visibility(
-                QueueUrl=self.sqs_queue_url,
-                ReceiptHandle=message.receipt_handle,
-                # we can only set integer timeouts
-                VisibilityTimeout=int(timeout),
-            )
+        timeout = self.backoff_policy.compute_backoff_timeout(message, visibility_timeout_seconds)
```
Comment (jessemyers) [inline microcosm_pubsub/consumer.py]: Delegate timeout calculation to backoff policy

File: microcosm_pubsub/consumer.py [hunk 4]
```diff
@@ -116,11 +121,19 @@ def configure_sqs_consumer(graph):
             region_name=region_name,
         )
 
+    backoff_policy_class = BackoffPolicy.choose_backoff_policy(
```
Comment (jessemyers) [inline microcosm_pubsub/consumer.py]: Choose the policy here.

**Comment by afallou** [comment]
+1

**Pull request closed without merging**